### ISSUE TITLE
Port extract_hog_features mask preparation to torch (remove per-face Python loop)

### ISSUE THREAD

**Issue by ljchang** [cosanlab/py-feat#293]
## Background

`feat/utils/image_operations.py::extract_hog_features` runs a per-face Python loop calling `extract_face_from_landmarks`. The HOG step itself was migrated to a batched torch \`HOGLayer\` in PR #283 — that part is fine. What still serializes is the **mask preparation** before HOG.

Per-face loop body, in `extract_face_from_landmarks` (`feat/utils/image_operations.py:109`):

1. \`align_face\` — builds a 3×3 affine matrix from 5 landmarks (eye centroids + nose + mouth corners) using Python \`float\` arithmetic over numpy, then warps via \`warp_affine\`.
2. \`scipy.spatial.ConvexHull(new_landmarks)\` — hull of the 68 aligned landmarks.
3. \`skimage.draw.grid_points_in_poly(shape=(112,112), verts=hull_verts)\` — point-in-polygon producing an HxW boolean mask.
4. \`mask_image\` — already torch.

For multi-face video the loop overhead compounds: 5 faces × ~30 fps × loop body ≈ real time. PR #292 cached \`HOGLayer\` itself, removing the per-call construction cost; that PR explicitly defers vectorizing the mask path because it's a real refactor with parity risk.

## Proposal: Option A — full torch port

- **Vectorize \`align_face\`** matrix construction across batch. The eye-centroid average, half-size bbox computation, and 3×3 matmul are all \`einsum\`-friendly across \`[N, 68, 2]\`.
- **Skip the dynamic convex hull entirely.** For 68 face landmarks, the hull is *deterministically* the outer contour: indices 0-16 (jaw) plus the top-of-forehead extension. Pre-bake the hull-vertex index pattern as a constant so we don't run a data-dependent algorithm on GPU.
- **Implement batched \`points_in_polygon\`** in torch via vectorized ray casting. \`[N, H*W, V_edges]\` crossing test → sum → odd-parity. ~30 lines.
- Result: single batched call, on-device, no CPU round trip per face.

## Parity risk

The xgboost and SVM AU classifiers were trained on features from the **exact** skimage-mask + skimage-HOG output. The HOG migration in PR #283 already eats ~5e-3 mean-abs drift, gated by \`test_extract_hog_features_matches_legacy_skimage_path\`. A torch-native mask path introduces another drift source. Two failure modes worth pinning before merge:

1. **Fixed-hull-index drift**: an extreme face pose may have a landmark interior to scipy's computed hull (e.g., closed eye, occluded jaw). The fixed-index pattern would clip differently. Need to validate across pose extremes.
2. **Ray-casting edge cases**: pixel exactly on a polygon edge is implementation-defined. skimage's \`grid_points_in_poly\` and a torch ray-cast won't necessarily agree on edge pixels. Could be ~0.1% pixel difference in the mask.

## Acceptance criteria

Before merging the torch port:

- [ ] **Test sweep**: 50+ real face crops covering pose extremes (profile, tilted, partially-occluded). Mostly synthetic + a sample from real video frames.
- [ ] **Per-face HOG parity**: \`assert_allclose\` between old (skimage-mask) and new (torch-mask) HOG features, atol ≤ 1e-2 (tighter than the 2e-2 PR #283 tolerance — the post-HOG bound).
- [ ] **Downstream AU stability**: AU predictions on \`multi_face.jpg\` and \`single_face.jpg\` should not flip categorical decisions vs the legacy path.
- [ ] **Benchmark**: report ms/face improvement on \`scripts/bench_detectors.py\` for both \`au_model='xgb'\` and \`au_model='svm'\` paths.

If parity holds within tolerance, replace the loop. If it doesn't, expose \`mask_strategy='legacy'|'torch'\` so the legacy path stays available for users on classifiers we can't retrain — and re-evaluate retraining a calibration on the new feature space as a v0.8 follow-up.

## Out of scope here

- Replacing scipy's \`ConvexHull\` with a torch convex-hull algorithm. Andrew's monotone chain is data-dependent and not a clean GPU fit. The fixed-pattern approach sidesteps that problem; if it doesn't hold parity, we revisit.
- Retraining the AU classifiers on torch-mask features (would close any drift gap but is a separate workstream — model retraining + skops publish + version bump).

## Pointers

- Loop site: \`feat/utils/image_operations.py::extract_hog_features\` (post-PR #292: takes a cached \`HOGLayer\`)
- Per-face helper: \`feat/utils/image_operations.py::extract_face_from_landmarks\`
- Existing parity test: \`feat/tests/test_extract_hog_features.py::test_extract_hog_features_matches_legacy_skimage_path\`
- HOG layer: \`feat/utils/image_operations.py::HOGLayer\` (already torch, already batched, OK to reuse)

**Comment by ljchang** [comment]
## Update: retraining is on the table

Owner is open to retraining xgboost + SVM AU classifiers if the speedup justifies the work. That meaningfully relaxes the design constraints captured above:

- The "preserve bitwise-or-near-bitwise feature parity" gate becomes optional — currently the hardest constraint.
- The fixed-hull-index trick stops being a parity-risk workaround and becomes a free design choice.
- Could go further: drop the convex-hull mask entirely and let HOG run on the aligned crop directly. Classifier learns to ignore background regions. Simpler code path, no polygon mask at all.
- Could replace the polygon mask with a precomputed fixed elliptical or rectangular face mask (constant tensor, broadcast-applied). Roughly the same masking effect at zero algorithmic cost.

## Two-stage plan

**Stage 1 (this issue, no retrain):** torch-native mask with fixed-hull pattern, parity gated against legacy skimage path. Ship if parity holds within ~1e-2 atol; gives current users a free speedup, no model swap.

**Stage 2 (separate workstream, retrain):** rebuild AU classifiers on whatever mask formulation produces best perf-per-line-of-code. Requires:
- Training data + labels (need to confirm what's still accessible)
- Calibration vs current classifier on a held-out set
- skops re-publish, version bump
- Migration story (do we keep both, deprecate v1, etc.)

Order matters: do Stage 1 first to set a baseline number for "torch port with parity," then measure how much additional speedup retraining unlocks. If Stage 1 already wins big, Stage 2 might not be worth it.

**Comment by ljchang** [comment]
## Constraint update: convex-hull mask is load-bearing for occlusion robustness

The convex-hull mask isn't cosmetic — it's likely **why** the AU classifier is robust to occlusions. Maintainer recall: this is reported in Cheong et al (Affective Science, 2023). The mask zeroes out pixels outside the face, which means the classifier never had occluding objects in its training distribution and can't fail-as-decoded on hair/hands/glasses crossing the cheek.

This sharpens the design constraints:

- **Stage 1 (parity-bound) is the safe path.** If the torch-native mask preserves the same masked-pixel set as the scipy/skimage version (within ~1px tolerance), occlusion robustness carries over by construction.
- **Dropping the mask entirely (a Stage 2 idea) is risky** even with retraining, because the training corpus probably wasn't curated for occlusion. Replicating that property without the mask would require an occlusion-augmented retrain, not just a swap of input crops.
- **The fixed-hull-index trick** (skip dynamic ConvexHull, use indices 0-16 jaw + forehead extension as a baked-in polygon) needs validation: does it produce the same masked region as scipy on extreme poses? If a closed eye or occluded jaw makes a landmark interior to scipy's hull, the fixed pattern would mask differently.

## Adjusted plan

Add an **occlusion-robustness regression test** to the acceptance criteria:

- [ ] Take real face crops, synthetically occlude a region (e.g., black square over the cheek, or a horizontal band across the eyes).
- [ ] Verify AU predictions on the occluded vs unoccluded face don't shift more than the legacy (skimage-mask) path's shift between the same pair.
- [ ] If the new mask path makes occlusion-shift LARGER, we have a regression — block merge.

This is a stricter gate than per-pixel mask parity and tests the actual property the system promises.

**Comment by ljchang** [comment]
## Stage 1 prototype landed: 4-5x speedup with bit-for-bit parity

Branch: \`experiment/torch-native-hog-mask\` (single commit on top of \`chore/v0.7-quick-wins-and-hog\`).

### Speedup (M5 MBP, MPS device)

| n_faces | Legacy | Skimage-batched (parity-safe) | Torch-batched (drift) |
|---|---|---|---|
| 5  | 14.6 ms | 6.6 ms (**2.2x**) | 6.3 ms (2.3x) |
| 20 | 55.6 ms | 14.8 ms (**3.8x**) | 10.5 ms (5.3x) |
| 50 | 140.9 ms | 31.2 ms (**4.5x**) | 19.8 ms (7.1x) |

CPU device shows ~5-10% improvement (loop wasn't the bottleneck there); MPS is where the per-face GPU↔CPU sync was eating ~98% of the time.

### Parity

End-to-end \`Detector(au_model='xgb', emotion_model='resmasknet').detect(single_face.jpg)\` AU values:

\`\`\`
Legacy vs batched-skimage path:
  shape: (1, 20)
  max |diff|:  0.00e+00
  mean |diff|: 0.00e+00
\`\`\`

**Bit-for-bit identical.** The skimage-batched path uses the exact \`scipy.spatial.ConvexHull\` + \`skimage.morphology.convex_hull.grid_points_in_poly\` calls with the same arguments as the legacy loop; the only thing batched is the alignment + final mask application + HOG, none of which alter the masked-pixel set.

### Architecture

Two new functions in \`feat/utils/_face_mask_torch.py\`:

- \`extract_hog_features_batched(faces, landmarks, hog_layer=None, mask_strategy='skimage')\` — drop-in replacement for the legacy \`extract_hog_features\` (same input contract: normalized [0, 1] landmarks).
- \`extract_faces_from_landmarks_batched(frames, landmarks_pix, face_size=112, mask_strategy='skimage')\` — lower-level wrapper that takes pixel-coord landmarks; called by the above.

The \`mask_strategy='torch'\` option uses an experimental on-device ray-cast (\`polygon_masks_batched\`). It's faster but currently disagrees with skimage on ~7 boundary pixels per face → 0.13 mean HOG drift. Gated for ablation only; would unlock the additional 1.4x on top of the skimage path if retraining is on the table (Stage 2).

### Tests

8 parity tests in \`feat/tests/test_face_mask_torch.py\`, all passing:
- align_faces_batched matches legacy align_face per-face
- polygon_masks_batched matches skimage on simple convex polygons
- extract_faces_from_landmarks_batched matches legacy loop
- **extract_hog_features_batched produces identical HOG features** (mean abs < 1e-4 vs legacy)

Plus the full existing test suite (66 tests covering the surrounding code) passes.

### Stage 2 (torch-batched, drift gated) — bigger picture

The 1.4x extra from torch-batched comes from removing the per-face skimage CPU call. To unlock it without retraining we'd need a torch-native polygon mask that exactly matches skimage's edge-inclusion semantics. Possible but real work (~50 lines of explicit edge rasterization). Probably not worth chasing if Stage 1 already meets the perf target — it's only 30-40% additional gain.

### Next decision points for owner

1. Merge Stage 1 into PR #292 (would tie this to the v0.7 quick-wins ship), or land as separate PR after #292?
2. Wire \`extract_hog_features_batched\` into Detector / MPDetector default path? Or behind a kwarg first?
3. Stage 2 retrain — pursue, or shelve given Stage 1 already wins big?

**Comment by ljchang** [comment]
Addressed in v0.7. The per-face Python loop in `extract_face_from_landmarks` is gone: `align_face` is now a batched torch port (`feat/utils/face_mask.py::align_faces_batched`) that vectorizes the affine construction + warp across all faces ([N,5,2] einsum), and the batched path is wired into the detector via `extract_hog_features_batched`. 

Note the convex-hull mask construction is *intentionally* kept on CPU (`scipy.spatial.ConvexHull`) for numerical exactness rather than approximated by a torch polygon kernel — see the rationale comment at the top of `face_mask.py`. The serializing affine loop this issue targeted is eliminated. Closing as done; can reopen if the remaining CPU hull step shows up as a bottleneck.
